File: kotlin/word-count/src/main/kotlin/WordCount.kt
object WordCount {
    fun phrase(phrase: String): Map<String, Int> {
        val words = phrase.filter { it.isWhitespace() || it.isLetter() }.split(' ')
        return words.groupBy { it }
                .map { it.key to it.value.size }
                .toMap()
    }

}

File: typescript/word-count/word-count.ts
export default class Words {
    count(sentence: string) {
        return sentence
            .toLowerCase()
            .split(/\s+/)
            .filter(Words.notEmpty)
            .reduce(Words.collectToMap, Words.emptyMap())
    }

    private static emptyMap(): Map<string, number> {
        return new Map<string, number>()
    }

    private static notEmpty(word: string): boolean {
        return word.length > 0
    }

    private static collectToMap(acc: Map<string, number>, word: string) {
        const count = acc.has(word) ? acc.get(word)! + 1 : 1
        acc.set(word, count)
        return acc
    }
}